waiter/src/waiter/discovery.clj

```diff
@@ -10,50 +10,42 @@
 ;;
 (ns waiter.discovery
   (:require [clojure.tools.logging :as log])
-  (:import org.apache.curator.x.discovery.ServiceDiscoveryBuilder
-           org.apache.curator.x.discovery.ServiceInstance
-           org.apache.curator.x.discovery.UriSpec
-           (org.apache.curator.x.discovery ServiceDiscovery ServiceCache)
-           (org.apache.curator.framework CuratorFramework)))
+  (:import (java.net Inet4Address)
+           (org.apache.curator.framework CuratorFramework)
+           (org.apache.curator.x.discovery ServiceCache ServiceDiscovery ServiceDiscoveryBuilder
+                                           ServiceInstance ServiceInstanceBuilder UriSpec)))
 
 (defn- ->service-instance
   [id svc-name {:keys [host port]}]
-  (let [builder (.. ServiceInstance
-                    builder
-                    (id id)
-                    (name svc-name)
-                    (uriSpec (UriSpec. "{scheme}://{address}:{port}/{endpoint}"))
-                    (port port))
-        builder' (if (= host "0.0.0.0")
-                   builder
-                   (.address builder host))]
-    (.build builder')))
+  (let [builder (-> (ServiceInstance/builder)
+                    (.id id)
+                    (.name svc-name)
+                    (.uriSpec (UriSpec. "{scheme}://{address}:{port}/{endpoint}"))
+                    (.port port)
+                    (.address (if (= host "0.0.0.0")
+                                (let [local-ips (ServiceInstanceBuilder/getAllLocalIPs)]
+                                  (->> (or (some #(when (instance? Inet4Address %) %) local-ips)
+                                           (first local-ips))
+                                       .getHostAddress))
```

Comment:
What will happen if none of them are `Inet4Address`?

Reply:
It will end up using the first `InetAddress` which is the current default behavior: https://github.com/barkbay/incubator-curator/blob/master/curator-x-discovery/src/main/java/org/apache/curator/x/discovery/ServiceInstance.java#L56

Reply:
Maybe we should just error out if there are no `Inet4Address`es

Reply:
Done.